framework/src/mediaplayer/mediaplayer.c

```diff
@@ -0,0 +1,581 @@
+{
+	if (m->state == MEDIA_STATE_PAUSED) {
+		m->state = MEDIA_STATE_PLAYING;
+		sem_post(&g_sem_playing);
+		return MEDIA_OK;
+	}
+	return MEDIA_ERROR;
+}
+
+void media_setvol(media_t *m, unsigned char vol)
+{
+	return;
+}
+
+bool media_isplaying(media_t *m)
+{
+	return (m->state == MEDIA_STATE_PLAYING);
+}
+
+media_result_t media_getvol(media_t *m)
+{
+	return MEDIA_ERROR;
+}
+
+media_result_t media_seek(media_t *m, int time_ms)
+{
+	return MEDIA_OK;
+}
+
+media_result_t media_setposition(media_t *m, int time_ms)
+{
+	return MEDIA_OK;
+}
+
+media_result_t media_getposition(media_t *m)
+{
+	return MEDIA_OK;
+}
+
+/***
+ * Functions run on threads
+ */
+
+static media_result_t read_wav_header(media_t *m)
+{
+	//struct pcm_config config;
+	struct wav_header_s header;
+	int ret;
+
+	ret = read(m->fd, &header, sizeof(header));
+	if (ret != sizeof(header)) {
+		return MEDIA_ERROR_UNKNOWN_FILE;
+	}
+
+	// Todo: Add header information to media_t
+
+	return MEDIA_OK;
+}
+
+media_t *media_open(char *path, media_op_t op, media_type_t type, media_play_type_t play_type)
+{
+	media_t *m = (media_t *)malloc(sizeof(media_t));
+	if (m == NULL) {
+		return NULL;
+	}
+
+	m->fd = -1;
+	m->op = op;
+	m->state = MEDIA_STATE_CREATED;
+	m->type = type;
+	m->play_type = play_type;
+
+	if (type == MEDIA_TYPE_WAV || type == MEDIA_TYPE_PCM) {
+		if (op == MEDIA_OP_PLAYBACK) {
+			if (play_type == MEDIA_PLAY_TYPE_FILE) {
+				m->fd = open(path, O_RDONLY);
+				if (m->fd < 0) {
+					free(m);
+					return NULL;
+				}
+
+				if (m->type == MEDIA_TYPE_WAV) {
+					read_wav_header(m);
+				}
+				m->state = MEDIA_STATE_PLAYING;
+				return m;
+			} else if (play_type == MEDIA_PLAY_TYPE_NETWORK) {
+				// TODO: Network streaming support
+				free(m);
+				return NULL;
+			}
+			free(m);
+			return NULL;
+
+		} else if (op == MEDIA_OP_RECORD) {
+			if (play_type == MEDIA_PLAY_TYPE_FILE) {
+				m->fd = open(path, O_RDWR | O_CREAT | O_TRUNC);
+				if (m->fd < 0) {
+					free(m);
+					return NULL;
+				}
+				m->state = MEDIA_STATE_RECORDING;
+				return m;
+			} else if (play_type == MEDIA_PLAY_TYPE_NETWORK) {
+				// TODO: Network streaming support
+				free(m);
+				return NULL;
+			}
+			free(m);
+			return NULL;
+		}
+	}
+	free(m);
+	return NULL;
+}
+
+void media_close(media_t *m)
+{
+	if (m == NULL) {
+		return;
+	}
+	if (m->fd >= 0) {
+		close(m->fd);
+	}
+	free(m);
+	return;
+}
+
+
+int on_media_state_playing(struct pcm *pcmout, media_t *m, char *buffer, unsigned int buffer_size)
+{
+	int readed;
+	int remain;
+	int ret;
+	readed = read(m->fd, buffer, buffer_size);
+	if (readed) {
+		remain = pcm_bytes_to_frames(pcmout, readed);
+		while (remain > 0) {
+			ret = pcm_writei(pcmout, buffer + readed - pcm_frames_to_bytes(pcmout, remain), remain);
+			if (ret <= 0) {
+				return MEDIA_ERROR_PCM_WRITE;
+			} else {
+				remain -= ret;
+			}
+		}
+	}
+
+	return readed;
+}
+
+int on_media_state_recording(struct pcm *pcmin, media_t *m, char *buffer, unsigned int buffer_size)
```

Comment:
Same as above

Reply:
In my opinion, here int return is needed especially in case of playing , because the num of bytes lets of know whether the playback has been completed or there is some part that is still left.

Reply:
Got it. you're right.

Reply:
I changed the return value.